README.md
About LnBlog
============
LnBlog is a weblog system that is written in PHP and stores data in text files.
It aims to provide a rich feature set with maximum flexibility while remaining
low on requirements.

Documentation
=============
Full documentation is available at the LnBlog website.  It can be viewed online 
at the following URL:  
https://www.skepticats.com/lnblog/documentation/

You may also download a copy of the documentation for local viewing at the 
following location:  
https://www.skepticats.com/lnblog/content/download/lnblog_docs.zip

Requirements
============
LnBlog needs a web server with PHP 7.0 or greater installed and the ability to
read and write to the local filesystem.  Support for the CURL, gettext, and 
either mime-magic or fileinfo extensions is helpful, but no non-standard 
extensions are required.

No database is needed and no particular web server is required.  LnBlog
is tested with Apache running on Windows and Linux.  Other web servers and
operating systems should work, but are not actively tested.

Installation
============
There are two ways to install: from the distribution ZIP file downloaded from
the LnBlog webpage (recommended) or from source, i.e. the cloned Git repository.

Installing from zip archive
---------------------------
To install the [ZIP file distribution](https://www.skepticats.com/lnblog/content/download/),
 simply extract the ZIP archive and upload 
the resulting folder to the publicly accessible portion of your web server.
After that, open a web browser go to the URL corresponding to that location.
This will start the graphical configuration process.  You will be prompted to 
confirm path information and create an initial user account, after which you will 
be taken to the administration page where you can create more users, create blogs,
and set other options.

Building from source
--------------------
To build LnBlog from source you will need the following installed on your system:
* Git
* PHP
* Composer
* Node.JS/NPM

To get the latest code and create an installable build, run the following
commands from the console:

```
git clone https://github.com/pageer/lnblog.git
cd lnblog
composer install --dev
vendor/bin/phing -Dversion='latest' build
```

This will create the directory `build/LnBlog-latest/`, which contains the same files
as the zip archive distribution described above.  You can copy that directory to 
the web-accessible portion of your web server and follow the installation instructions above.

Note that it is also possible to put the directory pulled from Git directly on your 
web server.  This makes it easy to pull updates to the code at will.  However, you will
have to manually handle changes to third-party dependencies.  For PHP dependencies, this means
you will have to run `composer install` to pull in updates.  For JavaScript dependencies,
you should create a build as described above and recursively copy the contents of the 
`build/third-party/` directory into `themes/default/` using something like this command:

```
cp -R build/third-party/* themes/default/
```

When pulling new changes, you may have to repeat these steps if any of the external 
dependencies change.  (Look for changes in `composer.lock`, `packages.json`, or `build.xml`.)

Upgrading
=========
To upgrade an existing LnBlog installation, extract the new ZIP archive 
(or create a new build from source) and upload the folder to your server.  You should 
then rename your old LnBLog directory to, e.g., LnBlog-old, and rename the new one 
in its place (i.e., give it the same name the old version had).  If your `userdata` directory
is contained inside the LnBlog directory, you should move it from the old version to the new one.
(Note: it is now recommended to keep your `userdata` directory alongside the code direcory,
not inside it.)

License
=======
LnBlog: A flexible file-base weblog
Copyright (C) 2005-2020 Peter A. Geer <pageer@skepticats.com>

This program is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software 
Foundation; either version 2 of the License, or (at your option) any later 
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY 
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this
program; if not, write to the Free Software Foundation, Inc., 675 Mass Ave, 
Cambridge, MA 02139, USA.

Credits
=======
Icons taken from Fam Fam Silk <http://www.famfamfam.com/lab/icons/silk/>


ChangeLog
04-Jul-2020: Version 2.1.0
Responsive layout (enhancement #11)
Drag-and-drop plugin loading (enhancement #4)
Auto-save for post editor (enhancement #1)
Paginated browsing (enhancement #5)
Asynchronous task runner (enhancement #6)
Deprecate file writing config (enhancement #13)
CSRF protection (enhancement #2)
Login attempt throttling (enhancement #3)
Make IP locking of session tokens optional (enhancement #10)
Password reset feature (enhancement #9)
Image and other links for drafts don't work in TinyMCE (bug #12)
Some file editor links don't work (bug #8)

21-Feb-2019: Version 2.0.0
Feature request #91: Add uploaded file awareness
Feature request #72: Add webmention support
Bug report #90: Plugin loading form broken
Bug report #89: XSL sheet doesn't load - cross-domain request blocked
Enhancement #58: Documentation cleanup
Bug report #82: Article and pingback settings not saved with draft
Enhancement #15: Separate admin pages
Bug report #20: Code uses include_path to find config and other files
Task #46: Switch to PHPUnit
Bug report #83: TinyMCE plugin override of markup type is confsing
Bug report #87: Make SSL work
Bug report #75: Edit post link goes to edit blog
Bug report #74: PCRE error in Blog::getPathFromEnvironment()

12-Aug-2015: Version 1.3.0
Bug report #73: Comments do not show homepage
Enhancement #52: Fix editor form layout
Enhancement #67: Clean up API stuff
Feature request #69: Admin interface to remove blogs
Bug report #71: Search plugin uses POST
Bug report #68: Getting blog in subdirectory doesn't work properly
Bug report #59: Need to be able to disable HTML editor on files
Enhancement #60: Search plugin providers
Feature request #61: Markdown text filter
Feature request #62: Add support for Composer
Feature request #63: Add a build system that invokes Composer

06-Jun-2014: version 1.2.0
Bug report 1: Missing security checks (pageer) - Fixed  
Enhancement 2: Allow non-sucky profile links 
Bug report 4: Post as article checkbox doesn't work (pageer) - Fixed  
Bug report 5: Publish/save buttons on published items (pageer) - Fixed  
Bug report 24: Warnings issued with E_STRICT (pageer) - Fixed  
Bug report 25: Pretty permalink URLs not recognized by pingback handler (pageer) - Fixed  
Bug report 26: Errors in tuxice theme (pageer) - Fixed  
Bug report 28: Relative upload URLs wrong with LBCode (pageer) - Fixed  
Enhancement 42: Link button in LBCode editor should have default (pageer) - Fixed  
Bug report 44: Multiple lists doesn't work (pageer) - Fixed  
Bug report 45: Unchecking "show in sidebar" does nothing (pageer) - Fixed  
Bug report 47: Add setting for time zone (pageer) - Fixed  
Enhancement 49: Add "disabled by default" setting for plugins (pageer) - Closed
Enhancement 51: New rich HTML editor (pageer) - Fixed  
Enhancement 53: Don't output for plugins with no data (pageer) - Fixed  
Bug report 54: Absolutizing of links in HTML mode broken (pageer) - Fixed  
Feature request 56: Need an about page (pageer) - Fixed  
Enhancement 57: TinyMCE for file editor (pageer) - Fixed  

12-Feb-2012: version 1.1.0
- 0000006: [Task] Pluggable text filters (pageer) - resolved.
- 0000031: [Improvement] Add "about me" sidebar plugin (pageer) - resolved.
- 0000035: [Task] Allow non-blog main page (pageer) - resolved.
- 0000036: [Defect] Calendar double display on month change (pageer) - resolved.
- 0000038: [Improvement] Add date to drafts (pageer) - resolved.
- 0000037: [Improvement] Port AJAX to jQuery (pageer) - resolved.
- 0000039: [Improvement] Busy indicator for AJAX preview (pageer) - resolved.
- 0000041: [Improvement] Ability to schedule publication (pageer) - resolved.
- 0000042: [Task] Use HTML5 (pageer) - resolved.

21-Jun-2011: version 1.0.0
- 0000017: [Improvement] Add an "insert link" button for uploads (pageer) - resolved.
- 0000021: [Defect] Breadcrumb plugin double entry name when on entry page (pageer) - resolved.
- 0000010: [Defect] Wrapper scripts should not rely on include_path (pageer) - resolved.
- 0000011: [Defect] Too many wrapper scripts (pageer) - resolved.
- 0000019: [Defect] Recent entries sidebar link broken when short tags is off (pageer) - resolved.
- 0000004: [Defect] Don't notify on friendly pingbacks (pageer) - resolved.
- 0000001: [Task] Dynamic file upload fields (pageer) - resolved.
- 0000018: [Defect] Article sidebar custom links edit link broken (pageer) - resolved.
- 0000013: [Vulnerability] Directory traversal vulnerability (pageer) - resolved.
- 0000009: [Defect] New blog PHP wrappers aren't created (pageer) - resolved.
- 0000008: [Defect] Warnings on fs setup (pageer) - resolved.

06-Nov-2007:
- Lots of fixes for things that broke in the process of refactoring.
- Changed "url" field in comment form, as it made the new entry display script thing that
the comment was an incoming pingback.  How long has THAT been broken?
- Added search path for userdata directory.  The blogconfig.php script will now look for userdata
in the LnBlog folder, or userdata, users, or profiles in the same directory as LnBlog.

09-Jul-2007:
- Lots of refactoring and added unit tests.
- Added data typing to the XML storage routines.

07-May-2007:
- Add separate template for entry permalinks.
- Allow only summary/abstract on blog front page.
- Added caching framework for output plugins.

16-Apr-2007:
- Made creating a blog on the document root work properly (as opposed to mostly working).
- Allow multiple files on in fileupload.php.

31-Mar-2007: version 0.9.0
- Fixed bad account detection in System::getUserList().

17-Mar-2007:
- Added PHP wrapping to solve cross-domain JavaScript problems.

13-Mar-2007:
- Fixed TrackbackValidator plugin to block trackbacks that have a non-URL in the
URL field.
- Fixed TrackBack receiving improperly showing a full page.

09-Mar-2007:
- Fixed Pingback errors by adding pingback() method to Trackback URI class.

lib/uri.php
<?php
/*
    LnBlog - A simple file-based weblog focused on design elegance.
    Copyright (C) 2005 Peter A. Geer <pageer@skepticats.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

# File: URI
# A library for handling URI generation and resolution.

# Function: create_uri_object
# Gets the appropriate URI generator class for a given object and type.  
# Basically, this the lazy dynamic programmer's version of a factory pattern.
#
# Parameters:
# object - The object for which to get a URI generator.  The class of the object
#          determines the class returned.  Note that this function supports 
#          one level of inheritance, i.e. if the class doesn't have a URI class
#          of its own, the parent class's is used.
# type   - Optional boolean that determines what kind of URI we're using, e.g.
#          querystring, htaccess, or wrapper scripts.  If not specified, then
#          the value of the URI_TYPE configuration constant is used.
#
# Returns:
# A URI generator object.  The exact object type is determined by the parameters
# to this function call.
function create_uri_object(&$object, $type=false) {
	$objtype = get_class($object);
	$partype = get_parent_class($object);
	if (!$type) $type = URI_TYPE;
	switch ($type) {
		case "hybrid":
			$type_ext = "URIHybrid";
			break;
		case "querystring":
			$type_ext = "URIQueryString";
			break;
		case "htaccess":
			$type_ext = "URIhtaccess";
			break;
		default:
			$type_ext = "URIWrapper";
	}
	
	# If the object class and URI type is not available, try the parent class.
	$creator = $objtype.$type_ext;
	if (! class_exists($creator)) $creator = $partype.$type_ext;
	
	# Neither the object nor parent class have the selected URI type, try
	# dropping back to the default URI type.
	$default_ext = "URIWrapper";
	if (! class_exists($creator)) $creator = $objtype.$default_ext;
	if (! class_exists($creator)) $creator = $partype.$default_ext;
	
	return new $creator($object);
}

###############################################################################
# Class: URI
# Represents a URI.  This auto-detects the URI settings from the environment, 
# such as the host, current path, etc., and allows the caller to override them
# to create other URIs.  URI strings should be built using this class.
###############################################################################

class URI extends LnBlogObject {
	function __construct() {
		$this->protocol = 'http';
		$this->host = isset($_SERVER['SERVER_NAME']) ? $_SERVER['SERVER_NAME'] : '';
		$this->port = isset($_SERVER['SERVER_PORT']) ? $_SERVER['SERVER_PORT'] : 80;
		$this->query_string = (! empty($_GET)) ? $_GET : array();
		$this->fragment = '';
		
		if (isset($_SERVER["REQUEST_URI"])) {
			$this->path = substr($_SERVER["REQUEST_URI"], 0, strpos($_SERVER["REQUEST_URI"], '?'));
		} else {
			$this->path = isset($_SERVER['PHP_SELF']) ? $_SERVER['PHP_SELF'] : '';
		}
		
	}
	
	function getQueryString($sep='&') {
		$qs = '';
		if (! empty($this->query_string)) {
			$pairs = array();
			foreach ($this->query_string as $key=>$val) $pairs[] ="$key=".urlencode($val);
			$qs = "?".implode($sep, $pairs);
		}
		return $qs;
	}
	
	function getFragment() {
		if ($this->fragment) return "#".$this->fragment;
		else return '';
	}
	
	function getPort() {
		if ($this->port != 80) return ":".$this->port;
		else return '';
	}
	
	# Method: get
	# Gets a string containingthe URI in the canonical format.
	function get() {
		return $this->protocol."://".
		       $this->host.
		       $this->getPort().
		       $this->path.
		       $this->getQueryString().
		       $this->getFragment();
	}
	
	# Method: getRelative
	# Gets a root-relative URI, i.e. no host, port, or protocol.
	function getRelative() {
		return $this->path.
		       $this->getQueryString().
		       $this->getFragment();
	}
	
	# Method: getPrintable
	# Like <get>, but returns a printable URI, i.e. with ampersands escaped.
	function getPrintable() {
		return $this->protocol."://".
			$this->host.
			$this->getPort().
			$this->path.
			$this->getQueryString('&amp;').
			$this->getFragment();
	}
	
	# Method: getRelativePrintable
	# Like <getPrintable>, but uses a relative URI.
	function getRelativePrintable() {
		return $this->path.
			$this->getQueryString('&amp;').
			$this->getFragment();
	}
	# Method: appendPath
	# Appends a path componend to the URI path, adding or removing terminal/beginning
	# slashes as appropriate.
	#
	# Parameters:
	# path - The path to append.
	function appendPath($path) {
		if ( substr($path, 0, 1) == "/") $path = substr($path, 1);
		if ( substr($this->path, -1) != "/") $this->path .= "/";
		$this->path .= $path;
	}
	
	# Method: clearQueryString
	# Empties the query string;
	function clearQueryString() {
		$this->query_string = array();
	}
	
	# Method: addQuery
	# Adds a parameter to the URI query string.
	#
	# Parameters:
	# key - The parameter name.
	# val - The value for that parameter.
	function addQuery($key, $val) {
		$this->query_string[$key] = $val;
	}
	
	# Method: secure
	# Set or get whether to use a secure, i.e. https, URL.
	function secure($val='') {
		if ($val === '') return $this->protocol == 'https';
		else {
			if ($this->protocol == 'https') $this->protocol = 'http';
			else $this->protocol = 'https';
			return $this->protocol;
		}
	}
}

class BlogURIWrapper extends URI {
	
	function BlogURIWrapper(&$blog) {
		parent::__construct();
	
		$this->object = $blog;
		
		if (defined('BLOG_ROOT_URL') && !$blog->skip_root) {
			$this->base_uri = BLOG_ROOT_URL;
		} else {
			$this->base_uri = localpath_to_uri($blog->home_path);
		}
		$url_bits = parse_url($this->base_uri);
		$url_components = array('protocol'=>'scheme',
		                        'host'=>'host', 'port'=>'port',
		                        'path'=>'path', 'username'=>'user',
		                        'password'=>'pass', 'path'=>'path',
		                        'query_string'=>'query',
		                        'fragment'=>'fragment');
		foreach ($url_components as $key=>$val) {
			if (isset($url_bits[$val])) {
				if ($key == 'query_string') {
					$this->{$key} = explode($url_bits[$val], '&');
				} else {
					$this->{$key} = $url_bits[$val];
				}
			}
		}
		
		$this->separator = "&amp;";
	}
	
	function setEscape($val) { $this->separator = $val ? "&amp;" : "&"; }
	
	function permalink() { return $this->base_uri; }
	function base() { return $this->base_uri; }
	function blog() { return $this->base_uri; }
	function page() { return $this->base_uri; }
	
	function articles() { return $this->base_uri.BLOG_ARTICLE_PATH."/"; }
	function entries() { return $this->base_uri.BLOG_ENTRY_PATH."/"; }
	function archives() { return $this->entries(); }
	function drafts() { return $this->base_uri.BLOG_DRAFT_PATH."/"; }
	function listdrafts() { return $this->drafts(); }
	
	function year($year) { 	
		return $this->base_uri.BLOG_ENTRY_PATH.sprintf("/%04d/", $year);
	}
	function listyear($year) { 	return $this->year($year); }
	
	function month($year, $month) { 
		return $this->base_uri.BLOG_ENTRY_PATH.sprintf("/%04d/%02d/", $year, $month);
	}
	function listmonth($year, $month) { return $this->month($year, $month); }
	
	function day($year, $month, $day) {
		return $this->base_uri.BLOG_ENTRY_PATH.
		       sprintf("/%04d/%02d/?day=%02d", $year, $month, $day);
	}
	function showday($year, $month, $day) { return $this->day($year, $month, $day); }
	
	function listall() { return $this->base_uri.BLOG_ENTRY_PATH."/?list=yes"; }
	
	function wrapper($action, $qs_arr=false) {
		$qs = '';
		if (is_array($qs_arr)) {
			foreach ($qs_arr as $key=>$val) {
				$qs .= $this->separator."$key=$val";
			}
		} elseif ($qs_arr) {
			$qs = $this->separator.$qs_arr;
		}
		return $this->base_uri."?action=$action".$qs;
	}
	
	function addentry() { return $this->wrapper("newentry"); }
	function addarticle() {
		return $this->wrapper("newentry", array("type"=>"article"));
	}
	function delentry($entryid) {
		return $this->wrapper("delentry", array('entry'=>$entryid));
	}
	function upload() { return $this->wrapper("upload"); }
	function edit() { return $this->wrapper("edit"); }
	function manage_reply() { return $this->wrapper("managereply"); }
	function manage_all() { return $this->wrapper("managereply"); }
	function manage_year($year) {
		return $this->wrapper("managereply", array('year'=>$year));
	}
	function manage_month($year, $month) {
		return $this->wrapper("managereply", array('year'=>$year, 'month'=>$month));
	}
	function login() {
		if (defined("LOGIN_HTTPS") && LOGIN_HTTPS) 
			$this->base_uri = localpath_to_uri($this->object->home_path, true, true);
		return $this->wrapper("login");
	}
	function logout() { return $this->wrapper("logout"); }
	function editfile($args) {
		if (! is_array($args)) $args = array('file'=> $args);
		return $this->wrapper("editfile", $args);
	}
	function edituser() { return $this->wrapper("useredit"); }
	function pluginconfig() { return $this->wrapper("plugins"); }
	function pluginload() { return $this->wrapper("pluginload"); }
	function tags($tag=false) { 
		if ($tag) {
			if (strpos($tag, "=")) {
				$data = explode("=", $tag);
				$tag = array($data[0]=>$data[1]);
			} else {
				$tag = array("tag"=>$tag);
			}
			return $this->wrapper("tags", $tag);
		} else {
			return $this->wrapper("tags"); 
		}
	}
	
	# These functions provide wrappers for cross-domain scripting.
	function script($script) { return $this->base_uri."?script=$script"; }
	function plugin($plug, $qs_arr=false) {
		$qs = "plugin=$plug";
		if (is_array($qs_arr)) {
			foreach ($qs_arr as $key=>$val) {
				$qs .= $this->separator."$key=$val";
			}
		}
		return $this->base_uri."?$qs"; 
	}

}

class BlogURIhtaccess extends BlogURIWrapper {

}

class BlogURIQueryString extends BlogURIWrapper {
	
}

class BlogURIHybrid extends BlogURIWrapper {
	
	function BlogURIHybrid(&$blog) {
		$this->object = $blog;
		$this->base_uri = localpath_to_uri($blog->home_path);
		$this->install_uri = INSTALL_ROOT_URL;
		$this->separator = "&amp;";
	}
	
	function addentry() { 
		$u = new URI();
		
		return make_uri($this->install_uri."pages/entryedit.php", 
	                    array("blog"=>$this->object->blogid));
	}
	function addarticle() {
		return make_uri($this->install_uri."pages/entryedit.php", 
	                    array("blog"=>$this->object->blogid,
		                      "type"=>"article"));
	}

}

class BlogEntryURIWrapper extends LnBlogObject {
	function BlogEntryURIWrapper(&$ent) {
		$this->object = $ent;
		$this->base_uri = localpath_to_uri(dirname($ent->file));
		$this->separator = "&amp;";
	}
	
	function setEscape($val) { $this->separator = $val ? "&amp;" : "&"; }
	
	function permalink() {
		$ent = $this->object;
		$pretty_file = $ent->calcPrettyPermalink();
		if ($pretty_file) {
			$pretty_file = mkpath(dirname($ent->localpath()),$pretty_file);
		}
		if ( file_exists($pretty_file) ) {
			# Check for duplicated entry subjects.
			$base_path = substr($pretty_file, 0, strlen($pretty_file)-5);
			$i = 2;
			if ( file_exists( $base_path.$i.".php") ) {
				$ret = $this->base_uri;
				while ( file_exists( $base_path.$i.".php" ) ) {
					$contents = file_get_contents($base_path.$i.".php");
					if (strpos($contents, basename(dirname($ent->file)))) {
						$ret = localpath_to_uri($base_path.$i.".php");
                        break;
					}
					$i++;
				}
			} else {
				$ret = localpath_to_uri($pretty_file);
			}
            if (defined('BLOG_ROOT_URL') && strpos($ret, $ent->getParent()->getURL()) !== 0) {
                $bad_root = localpath_to_uri($ent->getParent()->home_path);
                if (strpos($ret, $bad_root) === 0) {
                    $new_root = substr($ret, strlen($bad_root));
                    $ret = BLOG_ROOT_URL . $new_root;
                }
            }
            return $ret;
		} else {
			$pretty_file = $ent->calcPrettyPermalink(true);
			if ($pretty_file)
				$pretty_file = mkpath(dirname($ent->localpath()),$pretty_file);
			if ( file_exists($pretty_file) ) {
				return localpath_to_uri($pretty_file);
			} else {
				return $this->base_uri;;
			}
		}
	}
	
	function entry() { return $this->permalink(); }
	function page() { return $this->permalink(); } 
	
	function base() { return $this->base_uri; }
	function basepage() { return $this->base_uri."index.php"; } 
	
	function comment() { return $this->base_uri.ENTRY_COMMENT_DIR."/"; }
	function commentpage() { return $this->base_uri.ENTRY_COMMENT_DIR."/index.php"; }
	
	function send_tb() { return $this->base_uri.ENTRY_TRACKBACK_DIR."/?action=ping"; }
	function get_tb() { return $this->base_uri.ENTRY_TRACKBACK_DIR."/index.php"; }
	function trackback() { return $this->base_uri.ENTRY_TRACKBACK_DIR."/"; }
	function pingback() { return $this->base_uri.ENTRY_PINGBACK_DIR."/"; }
	function upload() {return $this->base_uri."/?action=upload"; }
	
	function edit() { return $this->base_uri."?action=edit"; }
	function editDraft($preview=false) {
		$b = NewBlog();
		$params = array('action'=>'edit', 'draft'=>$this->object->entryID());
		if ($preview) $params['preview'] = 'yes';
		return make_uri($b->uri('drafts'), $params, true, $this->separator);
	}
	
	function delete() {
		$ent =& $this->object;
		if ($ent->isDraft()) {
			$entry_type = 'draft';
		} else {
			$entry_type = is_a($ent, 'Article') ? 'article' : 'entry';
		}
		$qs_arr['blog'] = $ent->parentID();
		$qs_arr[$entry_type] = $ent->entryID();
		return make_uri(INSTALL_ROOT_URL."pages/delentry.php", $qs_arr);
	}
	
	function manage_reply() {
		return $this->base_uri."?action=managereplies";
	}
	function managereply() { return $this->manage_reply(); }
}

class BlogEntryURIHybrid extends BlogEntryURIWrapper {
	
	function BlogEntryURIHybrid(&$ent) {
		$this->object = $ent;
		$this->base_uri = localpath_to_uri(dirname($ent->file));
		$this->install_uri = INSTALL_ROOT_URL;
		$this->separator = "&amp;";
	}
	
	function edit() { 
		$b = NewBlog();
		return make_uri($b->uri("addentry"), 
		                array('entry'=>$this->object->entryID()));
	}
	
	function editDraft($preview=false) {
		$b = NewBlog();
		return make_uri($b->uri("addentry"), 
		                array('draft'=>$this->object->entryID()));
	}
}

class BlogCommentURIWrapper extends LnBlogObject {
	function BlogCommentURIWrapper(&$ent) {
		$this->object = $ent;
		$this->base_uri = localpath_to_uri(dirname($ent->file));
		$this->separator = "&amp;";
	}
	
	function setEscape($val) { $this->separator = $val ? "&amp;" : "&"; }
	
	function permalink() { return $this->base_uri."#".$this->object->getAnchor(); }
	function comment() { return $this->permalink(); }
	function delete() {
		$qs_arr = array('action' => 'delcomment');
		$parent = $this->object->getParent();
		$entry_type = is_a($this->object, 'Article') ? 'article' : 'entry';
		$qs_arr['blog'] = $parent->parentID();
		$qs_arr[$entry_type] = $parent->entryID();
		$qs_arr['delete'] = $this->object->getAnchor();
		return make_uri(false, $qs_arr);
	}
}

class TrackbackURIWrapper extends LnBlogObject {
	function TrackbackURIWrapper(&$ent) {
		$this->object = $ent;
		$this->base_uri = localpath_to_uri(dirname($ent->file));
		$this->separator = "&amp;";
	}
	
	function setEscape($val) { $this->separator = $val ? "&amp;" : "&"; }

	function trackback() { return $this->base_uri."#".$this->object->getAnchor(); }
	function pingback() { return $this->base_uri."#".$this->object->getAnchor(); }
	
	function permalink() { return $this->trackback(); }
	function delete() {
		$qs_arr = array('action' => 'delcomment');
		$parent = $this->object->getParent();
		$entry_type = is_a($this->object, 'Article') ? 'article' : 'entry';
		$qs_arr['blog'] = $parent->parentID();
		$qs_arr[$entry_type] = $parent->entryID();
		$qs_arr['delete'] = $this->object->getAnchor();
		return make_uri(false, $qs_arr);
	}
	
}
